README.md
<img src="vignettes/logo.png" align=right style="border:0;">

# cNorm

The package cNorm provides methods for generating non-parametric regression based continuous norms, as f. e. for psychometric test development, biometrics (biological and (e. g. physiological growth curves), and
screenings in the medical domain. It is based on the approach suggested by A. Lenhard et al. (2016).

## Approach

Conventional methods for producing test norms are often plagued with "jumps" or "gaps"
(i.e., discontinuities) in norm tables and low confidence for assessing extreme scores.
cNORM addresses these problems and also has the added advantage of not requiring
assumptions about the distribution of the raw data: The norm values are established from
raw data by modeling the latter ones as a function  of both percentile scores and an
explanatory variable (e.g., age) through Taylor polynomials. The method minimizes
bias arising from sampling and measurement error, while handling marked deviations from
normality – such as are commonplace in clinical samples. Contrary to parametric approaches, it does not rely on distribution assumptions of the initial norm data and is thus a very robust approach in generating norm tables.

## Example

Conducting the analysis consists of four steps:
1.  Data preparation
1.  Establishing the regression model and selecting the parameters
1.  Validating the model
1.  Generating norm tables and plotting the results

cNORM offers function for all of these steps, helps in selecting the best
fitting models and in generating the norm tables.

```{r example}
## basic example code for modeling the sample dataset
library(cNORM)
normData <- prepareData(elfe)
model <- bestModel(normData)

# Plot model
plotSubset(model)
plotPercentiles(normData, model)

# print norm table (for grade 3)
normTable(3, model)

# start vignette for a complete walk through
vignette(cNORM-Demo)
```
cNORM offers functions to choose the optimal model, both from a visual inspection of the 
percentiles, as well as by information criteria and model tests:

![](vignettes/plotSubset.png)

In this example, a Taylor polynomial with power k = 4 was computed in order to model a sample of the ELFE 1-6 reading comprehension test (sentence completion task; W. Lenhard & Schneider, 2006). In the plot, you can see the share of variance explained by the different models (with progressing number of predictors). Mallow's Cp (an AIC like measure) is used (BIC is available through the option `bic = TRUE`). The predefined adjusted R2 value of .99 is already reached with the third model and afterwards we only get minor improvements in adjusted R2. On the other hand, Cp rapidly declines afterwards, so model 3 seems to be a good candidate in terms of the relative information content per predictor and the captured information (adjusted R2). It is advisable to choose a model at the "elbow" in order to avoid over-fitting, but the solution should be tested for violations of model assumptions and the progression of the percentiles should be inspected visually, as well.


![](vignettes/plotPercentiles.png)

The predicted progression over age are displayed as lines and the manifest data as dots. Only three predictors were necessary to almost perfectly model the norm sample data with adjusted R2.

## Installation
Once it is on CRAN, cNORM can be installed via
```{r example}
install.packages("cNORM", dependencies = TRUE)
```

Until then, you can access the github development version via
```{r example}
install.packages("devtools")
devtools::install_github("WLenhard/cNORM")
library(cNORM)
```

The package is still under development. Please report errors. Suggestions for improvement are always welcome!

## Sample Data
The package includes data from two large test norming projects, namely ELFE 1-6 (Lenhard & Schneider, 2006) and the German adaption of the PPVT4 (A. Lenhard, Lenhard, Suggate & Seegerer, 2015), which can be used to run the analysis. Type `?elfe` or `?ppvt` to display information on the data sets.

## References
*   Lenhard, A., Lenhard, W., Suggate, S. & Segerer, R. (2016). A continuous solution to the norming problem. Assessment, Online first , 1-14. doi: 10.1177/1073191116656437
*   Lenhard, W., & Schneider, W. (2006). ELFE 1-6: Ein Leseverständnistest für Erst- bis Sechstklässler. Göttingen: Hogrefe. 


NEWS
TODO
====
- in general check parameter names for coherence (e. g. for the age / group / explanatory variable)
- Check rawTableQuick for usage with descending values
- predictNormValue: Fix computation of vectors



Version in 0.9.3
================
Date: 2018.08.20

Changes:
-------
- plotNormCurves enhanced (Thanks to Sebastian Gary)
- new function to plot semi parametric analyses via box cox power transformation: plotBoxCox



Version in 0.9.2
================
Date: 2018.08.18

Changes:
-------
- Addeditional dataset: vocabulary development (PPVT4)



Version in 0.9.1
================
Date: 2018.08.16

Changes:
-------
- Added predictRawBC and predictNormBC for computing norm and raw values based on the parametric
  box cox power function parameters
- New contributor: Sebastian Gary, welcome to the team!
- Missing raw variable definition in plotValues corrected



Version in 0.9.0
================
Date: 2018.08.14

Changes:
-------
- Box Cox power transformation for regression model at specific age: optional parametric modelling
  for non-paramteric regression model



Version in 0.8.9
================
Date: 2018.08.13

Changes:
-------
- Convenience method for selection best model added: 'printSubset'
- predictNormValue now accepts lists of values as well



Version in 0.8.8
================
Date: 2018.08.12

Changes:
-------
- parameter checks added
- new parameter 'descriptives' addad to rankByGroup and rankBySlidingWindow added
  to retrieve descriptive statistics for each observation
- improvements in the documentation
- errors in bestModel and plotPercentiles corrected, when variable names are not
  as in example sample



Version in 0.8.6
================
Date: 2018.08.11

Changes:
-------
- new function: 'rankBySlidingWindow' which can be used for data sets with continuous age variables
- error corrected for data being loaded from SPSS files
- improvements in the documentation



Version in 0.8.5
================
Date: 2018.08.06

Changes:
-------
- new function for simulating data



Version in 0.8.3
================
Date: 2018.08.03

Changes:
-------
- Code cleaning and formatting



Version in 0.8.2
================
Date: 2018.08.02

Changes:
-------
- new internal function: rawTableQuick for speeding up generating norm tables
  Still has to be checked for working with descending values.
  Works only, if model assumptions are valid



Version in 0.8.0
================
Date: 2018.08.01

Changes:
-------
- new function: rawTable allows creating norm tables with assignment of raw -> norm values
  solves inverse function of regression model with brute force



Version in 0.7.11
================
Date: 2018.07.31

Changes:
-------
- improved 'prepareData' function



Version in 0.7.10
================
Date: 2018.07.28

Changes:
-------
- Description for computePowers improved
- option in plotPercentile to use percentile scale or self defined c(mean, sd)



Version in 0.7.9
================
Date: 2018.07.28

Changes:
-------
- 'descend' parameter added to consistencyCheck and normTable
- dependency rColorBrewer removed; plotPercentiles changed accordingly
- latticeExtra moved to 'suggests'



Version in 0.7.8
================
Date: 2018.07.27

Changes:
-------
- 'descend' parameter added to consistencyCheck and normTable



Version in 0.7.7
================
Date: 2018.07.27

Changes:
-------
- Small changes to error messages in bestModel
- printing of min value in plotDerivate removed



Version in 0.7.6
================
Date: 2018.07.27

Changes:
-------
- parameter 'predictors' added to allow self defined regression functions,
  e. g. for the inclusion of other ranking parameters like sex
- 'type' parameter added to 'plotPercentile' to allow selection of quantile
  algorithm. Please consult help(quantile) for further information on 'type'



Version in 0.7.5
================
Date: 2018.07.26

Changes:
-------
- dependency dplyr removed: rankByOrder and plotPercentiles rewritten
- API-change: derivationPlot renamed to plotDerivate
- small changes to vignette and readme
- parameter "raw" added to rankByGroup to specify raw value variable
- rawVar and groupVar in plotPercentiles renamed to raw and group to
  make API more coherent


Version in 0.7.4
================
Date: 2018.07.25

Changes:
-------
- rankByOrder: ranking in descending order added



Version in 0.7.3
================
Date: 2018.07.25

Changes:
-------
- Additional ranking algorithms: Filliben, Levenbach, Yu & Huang; API changed to index
- scale can be specified as double vector with c(mean, sd)
- vignette updated accordingly



Version in 0.7.2
================
Date: 2018.07.24

Changes:
-------
- None. This is the first release


R/cNORM.R
#' cNORM: Continuous Norming with R
#'
#' The package provides methods for generating regression based continuous norms, as f. e.
#' for psychometric test development, biometrics (e. g. physiological growth curves), and
#' screenings in the medical domain. Contrary to parametric approaches, it does not rely on
#' distribution assumptions of the initial norm data and is thus a very robust approach in
#' generating norm tables.
#'
#' Conventional methods for producing test norms are often plagued with "jumps" or "gaps"
#' (i.e., discontinuities) in norm tables and low confidence for assessing extreme scores.
#' cNORM addresses these problems and also has the added advantage of not requiring
#' assumptions about the distribution of the raw data: The norm values are established from
#' raw data by modeling the latter ones as a function  of both percentile scores and an
#' explanatory variable (e.g., age). The method minimizes
#' bias arising from sampling and measurement error, while handling marked deviations from
#' normality - such as are commonplace in clinical samples.
#'
#' Conducting the analysis consists of four steps:
#' \enumerate{
#'   \item Data preparation
#'   \item Establishing the regression model and selecting the parameters
#'   \item Validating the model
#'   \item Generating norm tables and plotting the results
#' }
#'
#' cNORM offers function for all of these steps, helps in selecting the best
#' fitting models and generating the norm tables.
#'
#' @section Functions:
#' \enumerate{
#'   \item Data preparation: rankByGroup, computePowers
#'   \item Regression models: bestModel, regressionFunction, derive
#'   \item Model validation: checkConsistency, plotSubset, plotPercentiles,
#'   plotValues, derivationTable, plotDerative
#'   \item Norm tables: predictNormValue, predictRaw, normTable, getNormCurve, plotNormCurves
#' }
#' Example datasets with large cohorts are available for demonstration purposes ('elfe' and 'ppvt' sample data
#' from the references). Use \code{data <- prepareData(elfe)} or \code{data <- prepareData(ppvt)} to load and prepare example data
#' for the modeling. Use  \code{vignette(cNORM-Demo)} for a comprehensive
#' explanation for conducting  the modeling.
#'
#' @references
#' \enumerate{
#'   \item Lenhard, A., Lenhard, W., Suggate, S. & Segerer, R. (2016). A continuous solution to the norming problem. Assessment, Online first, 1-14. doi: 10.1177/1073191116656437
#'   \item Lenhard, A., Lenhard, W., Segerer, R. & Suggate, S. (2015). Peabody Picture Vocabulary Test - Revision IV (German Adaption). Frankfurt a. M.: Pearson Assessment.
#' }
#' @author Wolfgang Lenhard & Alexandra Lenhard
#' @keywords Psychometrics, Biometrics, Test Development, Regression Based Norming
#' @docType package
#' @name cNORM
NULL
